drivers/stepper/ti/drv8424.c

```diff
@@ -0,0 +1,775 @@
+	int32_t actual_position;
+	/** Target position. */;
+	int32_t target_position;
+	/** Whether the motor is currently moving. */
+	bool is_moving;
+	/** Whether we're in constant velocity mode. */
+	bool constant_velocity;
+	/** Which direction we're going if in constat velocity mode. */
+	enum stepper_direction constant_velocity_direction;
+	/** Event handler registered by user. */
+	stepper_event_callback_t event_callback;
+	/** Work item for offloading event callbacks. */
+	struct k_work event_callback_work;
+	/** Event handler user data. */
+	void *event_callback_user_data;
+	/** Message queue for passing data to event callback work item. */
+	struct k_msgq event_msgq;
+	/** Buffer for event message queue. */
+	char event_msgq_buffer[CONFIG_DRV8424_EVENT_QUEUE_LEN * sizeof(enum stepper_event)];
+	/** Whether the step signal is currently high or low. */
+	bool step_signal_high;
+	/** Struct containing counter top configuration. */
+	struct counter_top_cfg counter_top_cfg;
+};
+
+static void drv8424_user_callback_work_fn(struct k_work *work)
+{
+	int ret;
+
+	/* Get containing data struct */
+	struct drv8424_data *driver_data =
+		CONTAINER_OF(work, struct drv8424_data, event_callback_work);
+
+	/* Get next pending event (if any) from message queue */
+	enum stepper_event event;
+
+	ret = k_msgq_get(&driver_data->event_msgq, &event, K_NO_WAIT);
+	if (ret != 0) {
+		/* No pending events */
+		return;
+	}
+
+	/* Run the callback */
+	if (driver_data->event_callback != NULL) {
+		driver_data->event_callback(driver_data->dev, event,
+					    driver_data->event_callback_user_data);
+	}
+
+	/* If there are more pending events, resubmit this work item to handle them */
+	if (k_msgq_num_used_get(&driver_data->event_msgq) > 0) {
+		k_work_submit(work);
+	}
+}
+
+static int drv8424_schedule_user_callback(struct drv8424_data *data, enum stepper_event event)
+{
+	int ret;
+
+	/* Place event in message queue to be picked up by work item */
+	ret = k_msgq_put(&data->event_msgq, &event, K_NO_WAIT);
+	if (ret != 0) {
+		LOG_WRN("%s: Too many concurrent events, dropping event of type %d",
+			data->dev->name, event);
+		return -ENOBUFS;
+	}
+
+	/* Submit work item */
+	ret = k_work_submit(&data->event_callback_work);
+	if (ret < 0) {
+		LOG_ERR("%s: Failed to submit work item (error %d)", data->dev->name, ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static void drv8424_positioning_top_interrupt(const struct device *dev, void *user_data)
+{
+	struct drv8424_data *data = (struct drv8424_data *)user_data;
+	const struct drv8424_config *config = (struct drv8424_config *)data->dev->config;
+
+	if (!data->constant_velocity && data->actual_position == data->target_position) {
+		/* Check if target position is reached */
+		counter_stop(config->counter);
+		if (data->event_callback != NULL) {
+			/* Ignore return value since we can't do anything about it anyway */
+			drv8424_schedule_user_callback(data, STEPPER_EVENT_STEPS_COMPLETED);
```

Comment:
I think i understand the motivation behind using `msg_q` here, but i would have expected this not to be part of the driver, as soon as, one of this events is triggered, the driver just `notifies` via the stepper_event_callback_fn and then the queuing mechanism could happen in the application code, or in some sort of `motion_control` subsystem.

Reply:
The concern here is that the registered callback may not be safe to call in an interrupt context. Because we are triggering the `STEPPER_EVENT_STEPS_COMPLETED` event from within the interrupt generated by the counter, we move it into a (workqueue) thread context, where it presumably is safe to call (although it's still possible to construct deadlocks if one tries, since it runs on the system workqueue).

If the API specified that callbacks registered via `stepper_set_event_callback` have to be interrupt safe, this would not be necessary and we could call it directly.